lib/ansible/modules/file.py

```diff
@@ -469,7 +469,8 @@ def update_timestamp_for_file(path, mtime, atime, diff=None):
 
             set_time = (atime, mtime)
 
-        os.utime(b_path, set_time)
+        if not module.check_mode:
+            os.utime(b_path, set_time)
```

Comment:
this needs to check the time and return 'changed or not' information, both in our out of check mode .. to 'support checkmode'

Reply:
The diff/changed block is handled below, the os call wasn't actually used for that bit at all so that part still works.
```
# stat /tmp/ansible_mtime_test
  File: /tmp/ansible_mtime_test
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: fe01h/65025d    Inode: 792449      Links: 1
Access: (0664/-rw-rw-r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-02-08 16:19:49.737594350 +0000
Modify: 2022-02-08 16:19:38.213128892 +0000
Change: 2022-02-08 16:19:38.213128892 +0000
 Birth: 2022-02-07 14:15:03.137202224 +0000
# ansible-playbook test_mtime.yml --diff --check

PLAY [localhost] *********************************************************************************************

TASK [Gathering Facts] ***************************************************************************************
ok: [localhost]

TASK [Ansible mtime check mode test] *************************************************************************
--- before
+++ after
@@ -1,2 +1,2 @@
-mtime: 1644337178.2131288
+mtime: 1644329640.0
 path: /tmp/ansible_mtime_test

changed: [localhost]

PLAY RECAP ***************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

# stat /tmp/ansible_mtime_test
  File: /tmp/ansible_mtime_test
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: fe01h/65025d    Inode: 792449      Links: 1
Access: (0664/-rw-rw-r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-02-08 16:19:49.737594350 +0000
Modify: 2022-02-08 16:19:38.213128892 +0000
Change: 2022-02-08 16:19:38.213128892 +0000
 Birth: 2022-02-07 14:15:03.137202224 +0000
``` 

I can add in a changed block to the asserts in the test if you want?

Reply:
This is really not an issue with the PR but preexisting, currently the function always returns `true` but the callers don't check for differences before calling, so the first thing the function should do is check 'do we need a change?' after that is established, we can make the change (or not in check_mode) and resolve the diff info.

Reply:
something like this:
```
changed = False
...
if previous_mtime != mtime or previous_atime!= atime:
   changed = True
   if not module.check_mode:
         set_time(atime, mtime)
         os.utime(b_path, set_time)
....


return changed
```
We can also simplify the 'Sentinel' detection to skip the rest of the processing and return an empty diff

Reply:
There are some checks further up the function which do do the mtime comparisons and return False for normal runs, I don't know enough about how the Sentinel values are used to comment on that path. [file.py#L467-L468](https://github.com/ansible/ansible/blob/502fc1376750c2f566832adb73ec522b864e60d4/lib/ansible/modules/file.py#L467)
By the time it gets to this `os.utime` call, it has determined that the times are different.

Example double run to demo.
```
# ansible-playbook test_mtime.yml --diff

PLAY [localhost] *********************************************************************************************

TASK [Gathering Facts] ***************************************************************************************
ok: [localhost]

TASK [Ansible mtime check mode test] *************************************************************************
--- before
+++ after
@@ -1,2 +1,2 @@
-mtime: 1644340701.4436452
+mtime: 1644329640.0
 path: /tmp/ansible_mtime_test

changed: [localhost]

PLAY RECAP ***************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
# ansible-playbook test_mtime.yml --diff

PLAY [localhost] *********************************************************************************************

TASK [Gathering Facts] ***************************************************************************************
ok: [localhost]

TASK [Ansible mtime check mode test] *************************************************************************
ok: [localhost]

PLAY RECAP ***************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

Apologises if I'm misunderstanding or missing something.

Reply:
Based on the conversation, I've increased the covered of the integration test to cover the `.changed` and `.diff` blocks. I couldn't figure out why the diff always includes the path https://github.com/ansible/ansible/blob/502fc1376750c2f566832adb73ec522b864e60d4/lib/ansible/modules/file.py#L385-L388 and changing that seems like it could break something.

Reply:
makes it the same, should really not be the case .. but many things 'should not be the case' in this module ...

Reply:
The 2 paths do weirdly diff things, sentinel sets the `set_time` to `None` and the other does a `return False` (i had missed), so the changes as they are should be fine ... need to revamp this later so the flow is more obvious.